### DISCUSSION TITLE
Select stimulus based on timeline variable tag

### DISCUSSION THREAD

**Discussion by bettinanicole** [jspsych/jsPsych#855] [Q&A] [answered]
Hi there, 

I've searched around a bit and didn't seem to find something regarding my question, but I apologize in advance if it has been asked, there's a ton of feedback discussions and it's greatly appreciated!

I hope its an obvious/silly question, but I'm wondering whether there is a way of selecting specific stimuli from an array of timeline variables that matches a tag. I know there is a way of looking at past trials by looking at trial data such as with `jsPsych.data.get().filter()` but I'm wondering how to select future trials based on a condition. In the present task throughout one trial participants select a side and the side they choose dictates the proportion of A tagged trials or B tagged trials as you see in the example below. 

For example: 

`var stims = [
      { stimulus: "<img>i</img>",  tag: A, data: {xyz}}
      { stimulus: "<img>i</img>",  tag: A, data: {xyz}}
      { stimulus: "<img>s</img>",  tag: B, data: {xyz}}
      { stimulus: "<img>s</img>",  tag: B, data: {xyz}}
`
I'm using a custom plugin to have people choose a side and during the same trial, I'd like to conditionally select a stimulus using a tag (trial type). 

So in my custom plugin, you might see:

`If (left_chosen){
    80% of the time show tag A
    20% show tag B
 } else { 
    ...
 }`

In the index you would see something like: 

`var block = {
	timeline: [
	{
		type: "custom-plugin",
		stimulus:jsPsych.timelineVariable('stimulus'),
		data: jsPsych.timelineVariable('data'),
		tag: jsPsych.timelineVariable('tag'),
	},
         timeline_variables: stims,
}`

So in short, I'm wondering if there is a way of tagging stims and selecting them for future use?

I'm sure I can figure a way around this if not!

Thanks!

**Comment by jodeleeuw** [comment, marked as answer]
Welcome to the discussion forum @bettinanicole !

To make sure I'm understanding correctly, you want the decision about which stimulus to show to be made inside the plugin file?

With timeline variables, the trials on the timeline only get access to the particular variable set for that run through the timeline. So in your case the different images and their associated tags would end up creating separate trials, and you couldn't have access to both kinds of tags in the same run through a timeline. 

One way to solve this would be to have an A and a B tag for each trial. That may create issues with randomization. So depending on your response to my clarifying question we can brainstorm other ways around this.

**Reply by bettinanicole** [reply]
Thanks! 

Yes, I think it would have to be made within the plugin. So in the plugin, in the course of one trial participants would choose a side, and then once that side is chosen they are shown a stimulus, and the likelihood of seeing an A or B tagged stimulus depends on the side chosen which is defined in the plugin file. 

So, I think this answers my question in that the stimulus given for that trial is set at the beginning of the trial and cannot be changed in the middle of a trial from an event in the plugin when using timeline variables. But, with this in mind, I'm sure I can figure out ways of doing something similar to defining timeline variables but within the plugin. 

It's tricky to convey code and I'm sure even more so to understand what someone else is trying to do, but this already helps thank you!

**Reply by jodeleeuw** [reply]
> So, I think this answers my question in that the stimulus given for that trial is set at the beginning of the trial and cannot be changed in the middle of a trial from an event in the plugin when using timeline variables. 

Right. Once you are in the middle of a trial the parameters to that trial aren't meant to change. This is just due to the design of the library.

To offer a more concrete example, one thing you could do is set up your timeline variables like this

```js
var timeline_variables = [
	{stimulus_a: '<img src="a.png"></img>', stimulus_b: '<img src="b.png"></img>'},
	... etc ...
]
```

Then you could modify the plugin to accept two stimulus parameters, say `stimulus_a` and `stimulus_b`. In the plugin code you could then have an `if` statement like the one you proposed in your original question. You would render `trial.stimulus_a` or `trial.stimulus_b` depending on the choice made and random probability.

**Reply by bettinanicole** [reply]
Ah, great! Hadn't come to mind, that's exactly what I needed.

Thank you, 

Bettina
